File: library.cc
#include <fstream>
#include <sstream>
#include <stdexcept>
#include "library.hh"
#include "utils.hh"

namespace ndatk
{
  using namespace std;

  // Mass of neutron (U)
  static const double m_n = 1.00866491578; // "Nuclides and Isotopes" 16ed

  // Parse (e)xsdir
  void Library::parse(istream &s)
  {
    string line;
    enum states {START, AWR, DIR};
    states state = START;
    vector<string> fields;

    while (get_logical_line(s, line)) {
      if (starts_with_nocase(line, "DATAPATH")) {
        continue;             // TODO: add DATAPATH parse here
      } else if (starts_with_nocase(line, "ATOMIC WEIGH RATIOS")) {
        state = AWR;
      } else if (is_date(line)) {
        date = line;
      } else if (starts_with_nocase(line, "DIRECTORY")) {
        state = DIR;
      } else if (starts_with_nocase(line, "INCLUDE")) {
        fields = split(line);
        ifstream f1(fields[1].c_str());
        parse(f1);
        f1.close();
      } else if (state == AWR) {
        continue;               // ignore awr records
      } else if (state == DIR) {
        string id;
        DirectoryData d;
        istringstream r(line);
        r >> id >> d.awr >> d.name >> d.route >> d.type
          >> d.address >> d.tbl_len >> d.rcd_len >> d.epr
          >> d.temp >> d.ptable;
        directory.insert(Directory_map::value_type(id, d));
        order.push_back(id);
      }
    }
  }

  // Construct Library from data in stream
  Library::Library(istream& s)
  {
    Library::parse(s);
  }

  // Construct Library from data on filename
  Library::Library(const string filename)
  {
    ifstream s(filename.c_str());
    Library::parse(s);
    s.close();
    name = filename;
    info = "(e)xsdir";
  }

  // Return string value based on key
  string Library::get(string_val::key k) const
  {
    switch (k) {
    case string_val::NAME:
      return name;
      break;
    case string_val::DATE:
      return date;
      break;
    case string_val::INFO:
      return info;
      break;
    default:
      throw out_of_range("Key not found!");
    }
  }

  string Library::get(string_val_n::key k, int i) const
  {
    switch (k) {
    case string_val_n::ID:
      return order.at(i);
      break;
    default:
      throw out_of_range("Key not found!");
    }
  }

  int Library::get(int_val_x::key k, string id) const
  {
    DirectoryData d = map_at(directory, id);
    switch(k) {
    case int_val_x::ADDRESS:
      return d.address;
      break;
    case int_val_x::TBL_LEN:
      return d.tbl_len;
      break;
    case int_val_x::RCD_LEN:
      return d.rcd_len;
      break;
    case int_val_x::EPR:
      return d.epr;
      break;
    default:
      throw out_of_range("Key not found!");
    }
  }
      
  std::string Library::get(string_val_x::key k, string id) const
  {
    DirectoryData d = map_at(directory, id);
    switch(k) {
    case string_val_x::NAME:
      return d.name;
      break;
    case string_val_x::ROUTE:
      return d.route;
      break;
    case string_val_x::PTABLE:
      return d.ptable;
      break;
    default:
      throw out_of_range("Key not found!");
    }
  }

  double Library::get(float_val_x::key k, string id) const
  {
    DirectoryData d = map_at(directory, id);
    switch(k) {
    case float_val_x::AT_WGT:
      return d.awr * m_n;
      break;
    case float_val_x::AWR:
      return d.awr;
      break;
    case float_val_x::TEMP:
      return d.temp;
      break;
    default:
      throw out_of_range("Key not found!");
    }
  }
}

File: library.py
#! /usr/bin/env python
"""Library of Tables

Container for library of table IDs together with a map to ID metadata"""

__author__ = "Mark G. Gray"
__copyright__ = "Copyright 2013"
__credits__ = ["Mark G. Gray"]

__version__ = "1.0"
__maintainer__ = "Mark G. Gray"
__email__ = "gray@lanl.gov"
__status__ = "Development"

import re

            
class Library(object):
    # See MCNP6 Manual, Appendix F, Section II
    (AWR, FILE, ROUTE, TYPE, ADDRESS, TBL_LEN,
     RCD_LEN, EPR, TEMP, PTABLE) = range(10)
    TYPES = (float, str, str, int, int, int,
             int, int, float, str)
    def __init__(self, f):
        """Initialize Library from (e)xsdir records in f"""
        self.l = []             # List of ID in library
        self.d = {}             # Dictionary of ID meta-data
        self.parse(f)

    def parse(self, f):
        """Parse f for the following three sections:

        DATAPATH
        ATOMIC WEIGHT RATIOS
        DIRECTORY
        INCLUDE

        Update Library from DIRECTORY records"""

        date = re.compile(r'\d\d/\d\d/\d\d\d\d')
        state = ""
        for line in f:
            s = line.strip()           # Trim whitespace
            if len(s) > 0:             # Process nonblank nondate lines
                while s.endswith('+'): # Add continuation line
                    s = " ".join((s[:-1], f.next().strip()))
                # Process section keys
                key = s.lower()                # Canonicalize case
                if key.startswith("datapath"): # First section
                    if s.find('=') == -1:      # Optional '=' missing
                        fields = s.split()
                        self.datapath(fields[1])
                    else:
                        fields = s.spit('=')
                        self.datapath(fields[1].strip())
                elif key == "atomic weight ratios": # Second section
                    state = "awr"
                elif date.match(key):
                    self.date = key
                elif key == "directory": # Third section
                    state = "dir"
                elif key.startswith("include"): # Parse include file
                    fields = s.split()
                    f1 = open(fields[1])
                    self.parse(f1)
                    f1.close()
            # Process section data
                elif state == "awr":
                    pass            # Ignore awr records
                elif state == "dir":
                    fields = s.split()
                    zaid = fields[0]
                    self.l.append(zaid)
                    self.d[zaid] = tuple([t(v) for t, v in
                                          zip(Library.TYPES, fields[1:])])

    def make_library(self, pred):
        """Create a library object based on pred"""
        r = Library([])
        r.l = [z for z in self.l if pred(self, z)] # Filter zaids on pred
        r.d = self.d            # Reference Dictionary of ID meta-data
        return r

            
if __name__ == "__main__":
    import sys

    # Some test data from xsdir.lanl2006
    xsdir_dir = [
        "DIRECTORY",
        "1001.62c  0.999170 actia 0 1 1 10128 0 0 2.5301E-08",
        "1002.66c  1.996800 endf66a 0 1 2545 10270 0 0 2.5301E-08",
        "1003.69c  2.989596 t16_2003 0 1 1 11206 0 0 2.5301E-08",
        "2003.66c  2.989032 endf66a 0 1 6583 9679 0 0 2.5301E-08",
        "2004.62c  3.968219 actia 0 1 2545 5524 0 0 2.5301E-08"]

    d = Library(xsdir_dir)

    for zaid in d.l:
        print zaid, d.d[zaid][Library.AWR], d.d[zaid][Library.FILE], d.d[zaid][Library.TEMP]

    if len(sys.argv) == 2: 
        f = open(sys.argv[1])
        d2 = Library(f)

        print d2.date

        def pred(l, zaid):
            """Example predicate for library creation"""
            return l.d[zaid][Library.FILE].split('/')[0] == 'endf71x'

        endf71x = d2.make_library(pred)

        for za in endf71x.l:
            print za, endf71x.d[za][Library.FILE]
